test/micro/org/openjdk/bench/java/lang/StringDecode.java

```diff
@@ -70,90 +60,125 @@ public void setup() {
-             d\u00F6\u00F6\u00F6\u00F6\u00B6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6
-             e\u00F6\u00F6\u00F6\u00F6\u00F6\u00B6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6
-             f\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00B6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6
-             g\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00B6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6
-             h\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00B6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6
-             i\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00B6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6
-             j\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00B6\u00F6\u00F6\u00F6\u00F6\u00F6
-             k\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00B6\u00F6\u00F6\u00F6\u00F6
-             l\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00B6\u00F6\u00F6\u00F6
-             m\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00F6\u00B6\u00F6\u00F6
-            """.getBytes(charset);
+             Suspendisse potenti.""";
+    private static final String UTF16_STRING = "\uFF11".repeat(31);
+    private static final String LATIN1_STRING = "\u00B6".repeat(31);
+
+    @Setup
+    public void setup() {
+        charset = Charset.forName(charsetName);
+        asciiString = LOREM.substring(0, 32).getBytes(charset);
+        longAsciiString = LOREM.repeat(200).getBytes(charset);
+        utf16String = "UTF-\uFF11\uFF16 string".getBytes(charset);
+        longUtf16EndString = LOREM.repeat(4).concat(UTF16_STRING).getBytes(charset);
+        longUtf16StartString = UTF16_STRING.concat(LOREM.repeat(4)).getBytes(charset);
+        longUtf16OnlyString = UTF16_STRING.repeat(10).getBytes(charset);
+        latin1String = LATIN1_STRING.getBytes(charset);
+        longLatin1EndString = LOREM.repeat(4).concat(LATIN1_STRING).getBytes(charset);
+        longLatin1StartString = LATIN1_STRING.concat(LOREM.repeat(4)).getBytes(charset);
+        longLatin1OnlyString = LATIN1_STRING.repeat(10).getBytes(charset);
     }
 
     @Benchmark
-    public String decodeAsciiCharsetName() throws Exception {
-        return new String(asciiString, charsetName);
+    @CompilerControl(CompilerControl.Mode.DONT_INLINE)
+    public void decodeAsciiShort(Blackhole bh) throws Exception {
+        bh.consume(new String(asciiString, charset));
+        bh.consume(new String(asciiString, 0, 15, charset));
     }
 
     @Benchmark
-    public String decodeAscii() throws Exception {
-        return new String(asciiString, charset);
+    @CompilerControl(CompilerControl.Mode.DONT_INLINE)
+    public void decodeAsciiLong(Blackhole bh) throws Exception {
+        bh.consume(new String(longAsciiString, charset));
+        bh.consume(new String(longAsciiString, 0, 1024 + 31, charset));
```

Comment:
I imagine the 1024+31 addition gets compiled down, and is not executed during the test, right?

Reply:
Yes, adding two integer literals will be constant folded already by javac:
```
    public static void main(String...args) {
        int foo = 1024 + 31;
    }
```
javap -v output:
```      stack=1, locals=2, args_size=1
         0: sipush        1055
         3: istore_1
         4: return
```